### ISSUE TITLE
Vault CA provider doesn't handle fallback/retry on eventual consistency miss

### ISSUE THREAD

**Issue by krarey** [hashicorp/consul#14969]
#### Overview of the Issue

When configuring Consul to use Vault as its service mesh CA, the `/connect/ca/configuration` endpoint returns with an error if it encounters an HTTP 412 response from Vault.

This response is generated if a request reaches a Vault performance standby node which has not yet caught up with the leader. In these events, Vault's guidance is to enter a fallback/retry loop until the node catches up or the request is load-balanced to another node which can serve the request. [Vault Documentation](https://developer.hashicorp.com/vault/api-docs#412)

Specifically, I was able to trigger this event configuring a combined HCP Vault and HCP Consul environment via Terraform. At some point during CA configuration, a request from Consul to Vault was load-balanced to a node that was still catching up on its WAL and returned the following error:

```
│ Error: failed to set CA configuration: Unexpected response code: 500 (error configuring provider: Error making API request.
│ 
│ URL: GET https://krarey-zts-private-vault-ff4b2c13.ed765717.z1.hashicorp.cloud:8200/v1/auth/token/lookup-self
│ Code: 412. Errors:
│ 
│ * required index state not present)
```

#### Reproduction Steps

Unable to consistently reproduce this issue, as it is prone to Vault standby performance.

**Comment by jkirschner-hashicorp** [comment]
I strongly suspect that this would be fixed by upgrading the Vault API version used by Consul. The [latest Vault API code seems to automatically retry on a 412 code](https://github.com/hashicorp/vault/blob/06c0c6c8820669f6b70eb8e065ac6d2faf9dae20/api/client.go#L1735-L1748). That code seems to have been [introduced](https://github.com/hashicorp/vault/commit/22b486b651b8956d32fb24e77cef4050df7094b6) in Feb 2021, which means it was probably released in [Vault API version ~1.1.0 in March 2021](https://pkg.go.dev/github.com/hashicorp/vault/api?tab=versions). Consul is [using Vault API version 1.0.5](https://github.com/hashicorp/consul/blob/c7ef04c5979dbc311ff3c67b7bf3028a93e8b0f1/go.mod#L64), which lacks the automatic retry on a 412 code.

Therefore, I suspect that if we upgraded the Vault API version (https://github.com/hashicorp/consul/issues/12613), we could close this.

**Comment by jkirschner-hashicorp** [comment]
Closed by https://github.com/hashicorp/consul/pull/15970.

That PR updated the Vault API to 1.8.2.
